src/arithmetic/agg_funcs.c

```diff
@@ -454,15 +508,84 @@ AggCtx *Agg_CollectFunc() {
 
 //------------------------------------------------------------------------
 
+typedef struct {
+	SIValue list;
+} __agg_collectDistinctCtx;
+
+int __agg_collectDistinctStep(AggCtx *ctx, SIValue *argv, int argc) {
+	// convert multiple values to array
+
+	assert(argc >= 0);
+	__agg_collectDistinctCtx *ac = Agg_FuncCtx(ctx);
+
+	// TODO: can this be migrated to Agg_CollectDistinctFunc?
+	if(ac->list.type == T_NULL) ac->list = SI_Array(argc);
+	for(int i = 0; i < argc; i ++) {
+		SIValue value = argv[i];
+		if(value.type == T_NULL) continue;
+		SIArray_Append(&ac->list, value);
+	}
+	return AGG_OK;
+}
+
+int __agg_collectDistinctReduceNext(AggCtx *ctx) {
+	__agg_collectDistinctCtx *ac = Agg_FuncCtx(ctx);
+
+	if(SIValue_IsNull(ac->list)) {
+		Agg_SetResult(ctx, ac->list);
+		return AGG_OK;
+	}
+
+	uint32_t item_count = SIArray_Length(ac->list);
+	if(item_count <= 1) {
+		Agg_SetResult(ctx, ac->list);
+		return AGG_OK;
+	}
+
+	SIValue *items = ac->list.array;
+	SIValue distinct_array = SI_Array(item_count);
+
+	/* Sort array, don't count duplicates */
+	QSORT(SIValue, items, item_count, SI_LT);
```

Comment:
How about consider the usage of rax as a set, and calling `raxFind` as a set membership validation.
You can call n times `raxFind` which is n*(the cost of hash). If the result for the current value is `RAX_NOT_FOUND` add the current value both to the distinct array and the rax.
In the current implementation, you'll have n*logn which is the cost of sort + n times comparison for the distinct filter.
Maybe another benefit is the removal of the implicit ordering in the result array.